File: Códigos das Implementações/Código em Java/MergeSortAlgorithm.java
// IMPLEMENTAÇÃO MERGE SORT EM JAVA
import java.io.*;
import java.util.*;

public class MergeSortAlgorithm {
    private List<Integer> array = new ArrayList<>();
    private String arq;
    private String arq2;

    public MergeSortAlgorithm(String arqE, String arqS) {
        this.arq = arqE;
        this.arq2 = arqS;
    }

    private void merge(List<Integer> v, int inicio, int meio, int fim) {
        List<Integer> temp = new ArrayList<>();
        int p1 = inicio, p2 = meio + 1;

        while (p1 <= meio && p2 <= fim) {
            if (v.get(p1) <= v.get(p2)) {
                temp.add(v.get(p1++));
            } else {
                temp.add(v.get(p2++));
            }
        }

        while (p1 <= meio) {
            temp.add(v.get(p1++));
        }

        while (p2 <= fim) {
            temp.add(v.get(p2++));
        }

        for (int i = 0; i < temp.size(); i++) {
            v.set(inicio + i, temp.get(i));
        }
    }

    private void mergeSort(List<Integer> v, int inicio, int fim) {
        int meio;
        if (inicio < fim) {
            meio = (inicio + fim) / 2;
            mergeSort(v, inicio, meio);
            mergeSort(v, meio + 1, fim);
            merge(v, inicio, meio, fim);
        }
    }

    private void definirArray(int n) {
        try (Scanner scanner = new Scanner(new File(arq))) {
            array.clear(); 
            while (scanner.hasNextInt()) {
                array.add(scanner.nextInt());
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        if (array.size() > n) {
            array = array.subList(0, n);
        }
    }

    private void salvarTempo(int n, double timeTaken) {
    try (PrintWriter writer = new PrintWriter(new FileWriter(arq2, true))) {
        writer.print("\nJava," + n + "," + String.format(Locale.US, "%.10f", timeTaken) + "," + arq);
    } catch (IOException e) {
        e.printStackTrace();
    }
}

public void run(int n) {
    definirArray(n);

    long startTime = System.nanoTime();
    mergeSort(array, 0, array.size() - 1);
    long endTime = System.nanoTime();
    double timeTaken = (endTime - startTime) / 1e9;

    salvarTempo(n, timeTaken);
}

}

File: src/Códigos das Implementações/Código em JavaScript/MergeSortAlgorithm.js
const fs = require('fs');
const readline = require('readline');

class MergeSortAlgorithm {
    constructor(arqE, arqS) {
        this.array = [];
        this.arq = arqE;
        this.arq2 = arqS;
    }

    merge(v, inicio, meio, fim) {
        let temp = [];
        let p1 = inicio, p2 = meio + 1;

        while (p1 <= meio && p2 <= fim) {
            if (v[p1] <= v[p2]) {
                temp.push(v[p1++]);
            } else {
                temp.push(v[p2++]);
            }
        }

        while (p1 <= meio) {
            temp.push(v[p1++]);
        }

        while (p2 <= fim) {
            temp.push(v[p2++]);
        }

        for (let i = 0; i < temp.length; i++) {
            v[inicio + i] = temp[i];
        }
    }

    mergeSort(v, inicio, fim) {
        let meio;
        if (inicio < fim) {
            meio = Math.floor((inicio + fim) / 2);
            this.mergeSort(v, inicio, meio);
            this.mergeSort(v, meio + 1, fim);
            this.merge(v, inicio, meio, fim);
        }
    }

    definirArray(n) {
        return new Promise((resolve, reject) => {
            fs.readFile(this.arq, 'utf8', (err, data) => {
                if (err) {
                    return reject(err);
                }

                this.array = data.split(/\s+/).map(Number);
                if (this.array.length > n) {
                    this.array = this.array.slice(0, n);
                }
                resolve();
            });
        });
    }

    salvarTempo(n, timeTaken) {
        const data = `\nJavaScript,${n},${timeTaken},${this.arq}`;
        fs.writeFileSync(this.arq2, data, { flag: 'a' });
    }

    printArray() {
        console.log(this.array);
    }

    async run(n) {
        await this.definirArray(n);

        const startTime = process.hrtime();
        this.mergeSort(this.array, 0, this.array.length - 1);
        const endTime = process.hrtime(startTime);
        const timeTaken = endTime[0] + endTime[1] / 1e9;

        this.salvarTempo(n, timeTaken);
    }
}

module.exports = MergeSortAlgorithm;
